### PR TITLE
Allow delete multiple items - Group pages

### PR DESCRIPTION
**Pull request by rdwebdesign**
**By submitting this pull request, I confirm the following:**

- [x] I have read and understood the [contributors guide](https://github.com/pi-hole/AdminLTE/blob/master/CONTRIBUTING.md), as well as this entire template.
- [x] I have made only one major change in my proposed changes.
- [x] I have commented my proposed changes within the code.
- [x] I have tested my proposed changes.
- [x] I am willing to help maintain this change if there are issues with it later.
- [x] I give this submission freely and claim no ownership.
- [x] It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)
- [x] I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))
- [x] I have Signed Off all commits. ([`git commit --signoff`](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff))

---

**What does this PR aim to accomplish?:**

Add controls to the every groups pages to allow selection and deletion of multiple items at once.

**How does this PR accomplish the above?:**

This PR adds check-boxes, buttons and control functions to allow delete multiple items.

**What documentation changes (if any) are needed to support this PR?:**
none

### CODE REVIEW

File: scripts/pi-hole/js/groups-adlists.js [hunk 1]
```diff
@@ -117,24 +117,32 @@ function initTable() {
     order: [[0, "asc"]],
     columns: [
       { data: "id", visible: false },
-      { data: "status", searchable: false, class: "details-control" },
-      { data: "address" },
+      { data: null, visible: true, orderable: false, width: "15px" },
+      { data: "status", searchable: false, orderable: false, class: "details-control" },
```
Comment (yubiuser) [inline scripts/pi-hole/js/groups-adlists.js]: Why did you prevent ordering here?
```suggestion
      { data: "status", searchable: false, class: "details-control" },
```

File: scripts/pi-hole/js/groups-adlists.js [hunk 2]
```diff
@@ -117,24 +117,32 @@ function initTable() {
...
+          return "";
+        },
       },
       {
         targets: "_all",
         render: $.fn.dataTable.render.text(),
       },
     ],
     drawCallback: function () {
+      // Hide buttons if all messages were deleted
```
Comment (yubiuser) [inline scripts/pi-hole/js/groups-adlists.js]:
```suggestion
      // Hide buttons if all adlists were deleted
```

File: scripts/pi-hole/js/groups-clients.js [hunk 3]
```diff
@@ -92,18 +92,31 @@ function initTable() {
...
+          return "";
+        },
+      },
       {
         targets: "_all",
         render: $.fn.dataTable.render.text(),
       },
     ],
     drawCallback: function () {
+      // Hide buttons if all messages were deleted
```
Comment (yubiuser) [inline scripts/pi-hole/js/groups-clients.js]:
```suggestion
      // Hide buttons if all clients were deleted
```

File: scripts/pi-hole/js/groups-domains.js [hunk 4]
```diff
@@ -77,12 +78,24 @@ function initTable() {
...
+          return "";
+        },
+      },
       {
         targets: "_all",
         render: $.fn.dataTable.render.text(),
       },
     ],
     drawCallback: function () {
+      // Hide buttons if all messages were deleted
```
Comment (yubiuser) [inline scripts/pi-hole/js/groups-domains.js]:
```suggestion
      // Hide buttons if all domains were deleted
```

File: scripts/pi-hole/js/groups.js [hunk 5]
```diff
@@ -22,18 +22,31 @@ $(function () {
...
+          return "";
+        },
+      },
       {
         targets: "_all",
         render: $.fn.dataTable.render.text(),
       },
     ],
     drawCallback: function () {
+      // Hide buttons if all messages were deleted
```
Comment (yubiuser) [inline scripts/pi-hole/js/groups.js]:
```suggestion
      // Hide buttons if all groups were deleted
```

File: scripts/pi-hole/js/groups.js [hunk 6]
```diff
@@ -106,7 +166,7 @@ $(function () {
       }
 
       // Reset visibility of ID column
-      data.columns[0].visible = false;
+      data.columns[1].visible = false;
```
Comment (yubiuser) [inline scripts/pi-hole/js/groups.js]:
```suggestion
      data.columns[0].visible = false;
```

ID should still be column `0`, or?

File: scripts/pi-hole/js/groups-adlists.js (hunk 1 above)
Comment (rdwebdesign) [reply scripts/pi-hole/js/groups-adlists.js]: sorting-padding

I was testing the column without the sorting and forgot to re-enabled it.

File: scripts/pi-hole/js/groups.js (hunk 6 above)
Comment (rdwebdesign) [reply scripts/pi-hole/js/groups.js]: Investigating.

Comment (rdwebdesign) [reply scripts/pi-hole/js/groups.js]: Yes... It should be `0`.

**Review by yubiuser — COMMENTED** [review]
I've found a likely expected, but weird behavior. If you select "all" but then add an other group/domain/client/adlist it leaves the "select toggle" in a weird undefined stage. I would expect that it deselects everything if new items are added to the tables

Peek 2022-05-06 21-48

**Comment by rdwebdesign** [comment]
I notice the `response` is successful and data is added correctly to the database.
The error happens on line 445, when javascript reloads the data:

https://github.com/pi-hole/AdminLTE/blob/89fdc7ed1cf80e62aa0d4e5ac4cbf239aa1d1b8f/scripts/pi-hole/js/groups-clients.js#L442-L446

File: scripts/pi-hole/js/groups-adlists.js [hunk 7]
```diff
@@ -117,24 +117,32 @@ function initTable() {
     order: [[0, "asc"]],
     columns: [
       { data: "id", visible: false },
+      { data: null, visible: true, orderable: false, width: "15px" },
       { data: "status", searchable: false, class: "details-control" },
-      { data: "address" },
+      { data: "address", orderable: false },
```
Comment (yubiuser) [inline scripts/pi-hole/js/groups-adlists.js]: I know it was like not orderable before, but I don't see a reason why it should not be orderable...

```suggestion
      { data: "address" },
```

File: scripts/pi-hole/js/groups-clients.js [hunk 8]
```diff
@@ -92,18 +92,31 @@ function initTable() {
     order: [[0, "asc"]],
     columns: [
       { data: "id", visible: false },
+      { data: null, visible: true, width: "15px" },
```
Comment (yubiuser) [inline scripts/pi-hole/js/groups-clients.js]:
```suggestion
      { data: null, visible: true, orderable: false, width: "15px" },
```

File: scripts/pi-hole/js/groups.js [hunk 9]
```diff
@@ -22,18 +22,31 @@ $(function () {
     order: [[0, "asc"]],
     columns: [
       { data: "id", visible: false },
+      { data: null, visible: true, width: "15px" },
```
Comment (yubiuser) [inline scripts/pi-hole/js/groups.js]:
```suggestion
      { data: null, visible: true, orderable: false, width: "15px" },
```

**Review by yubiuser — COMMENTED** [review]
Just for consistency. You used twice `orderable: false` and twice not. You could also just remove it twice.

**Review by yubiuser — COMMENTED** [review]
In `groups.php` you have several places where you have an exploit prevention. But only for deleting adlists you added

```
        // Accept only an array
        $ids = json_decode($_POST['id']);
        if (!is_array($ids)) {
           throw new Exception('Invalid payload: id is not an array');
        }
        
```

Maybe we should also check in case we want to delete clients, domains or groups.

File: scripts/pi-hole/js/groups-adlists.js (hunk 7 above)
Comment (rdwebdesign) [reply scripts/pi-hole/js/groups-adlists.js]: Changed

**Review by yubiuser — APPROVED** [review]

**Pull request merged**